article.md

```diff
@@ -0,0 +1,333 @@
+// value has to be a 16 byte word
+// prefix length = PREFIX_BITS
+func add_prefix{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}(value : felt, prefix : felt) -> (
+        result : felt, overflow: felt){
+    let shifted_prefix = prefix * 2 ** 128;
+    // with_prefix is 18 bytes long
+    let with_prefix = shifted_prefix + value;
+    let overflow_mask = 2 ** 16 - 1;
+    let (overflow) = bitwise_and(with_prefix, overflow_mask);
+    let result = (with_prefix - overflow) / 2 ** 16;
+    return (result, overflow);
+}
+```
+We're again operating on 128 bit values, so we need to shift everything to fit our prefix.
+
+```cairo
+    let prefix = PREFIX;
+    let (w0, prefix) = add_prefix(domain_hash.high, prefix);
+    let (w1, prefix) = add_prefix(domain_hash.low, prefix);
+    let (w2, prefix) = add_prefix(data_hash.high, prefix);
+    let (w3, overflow) = add_prefix(data_hash.low, prefix);
+```
+
+We can't use the same high-level keccak implementation as before (`keccak_uint256s_bigend`), since it assumes that our input in bytes is a multiple of 32, which it is not. Because of the 2 overflow bytes, our input is 66 bytes long, and thus `keccak_uint256s_bigend` will assume that we're also counting all the leading zeros of the last value, making our input 96 bytes long, which would result in incorrect hashes. The solution is to use the low-end keccak function - `keccak_bigend`, which unfortunately takes the input in the little-endian format, forcing us to change our variables to it. Note that the `bigend` in the name of the function refers to the output format, not the input format.
+
+```cairo
+// Splits 128bit big endian int to 2 64bit small endian ints
+func split_to_64bit_small_endian_words{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}(value: felt) -> (high: felt, low: felt){
+    let (temp) = word_reverse_endian(value);
+    let (high, low) = unsigned_div_rem(temp, DIVISOR);
+    return (high, low);
+}
+// Changes input format to array of 64bit values in little endian format
+func populate_array_little_end{range_check_ptr, bitwise_ptr : BitwiseBuiltin*}(w0: felt, w1: felt, w2: felt, w3: felt, overflow: felt, signable_bytes : felt*){
+    
+    let (high, low) = split_to_64bit_small_endian_words(w0);
+    assert signable_bytes[0] = low;
+    assert signable_bytes[1] = high;
+
+    let (high, low) = split_to_64bit_small_endian_words(w1);
+    assert signable_bytes[2] = low;
+    assert signable_bytes[3] = high;
+
+    let (high, low) = split_to_64bit_small_endian_words(w2);
+    assert signable_bytes[4] = low;
+    assert signable_bytes[5] = high;
+
+    let (high, low) = split_to_64bit_small_endian_words(w3);
+    assert signable_bytes[6] = low;
+    assert signable_bytes[7] = high;
+
+    // Overflow is only 2 bytes long
+    let (temp) = word_reverse_endian(overflow);
+    let (value8, trash) = unsigned_div_rem(temp, 2**112);
+    assert signable_bytes[8] = value8;
+
+    return ();
+}
+```
+Now we can finish our `get_hash` function.
```

Comment:
I think it would be better to name it `get_eip712_hash`, so it is clearly visible.